expression/builtin_string.go

```diff
@@ -922,9 +922,36 @@ func (c *convertFunctionClass) getFunction(ctx sessionctx.Context, args []Expres
 		return nil, err
 	}
 	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETString, types.ETString, types.ETString)
-	// TODO: issue #4436: The second parameter should be a constant.
-	// TODO: issue #4474: Charset supported by TiDB and MySQL is not the same.
-	// TODO: Fix #4436 && #4474, set the correct charset and flag of `bf.tp`.
+
+	// Validate charset before evaluating conversion, to make charset and flag of `bf.tp` correct
+	// even if it's not evaluated. (#4436)
+	transcodingName, isNull, err := args[1].EvalString(ctx, chunk.Row{})
```

Comment:
in the stage of `getFunction()`, `EvalXXX()` functions are not supposed to be called. Because the former is in the plan building stage while the later is in the execution stage.

Reply:
On my local MySQL 5.7.25, I think `args[1]` is also constrained by parser to be a constant string. Here's my observation:
```
mysql> SHOW VARIABLES LIKE "%version%";
+-------------------------+------------------------------+
| Variable_name           | Value                        |
+-------------------------+------------------------------+
| innodb_version          | 5.7.25                       |
| protocol_version        | 10                           |
| slave_type_conversions  |                              |
| tls_version             | TLSv1,TLSv1.1                |
| version                 | 5.7.25                       |
| version_comment         | MySQL Community Server (GPL) |
| version_compile_machine | x86_64                       |
| version_compile_os      | Linux                        |
+-------------------------+------------------------------+
8 rows in set (0.00 sec)

mysql> drop table if exists t;
Query OK, 0 rows affected (0.02 sec)

mysql> create table t (a varchar(20));
Query OK, 0 rows affected (0.06 sec)

mysql> insert t values ('utf8');
Query OK, 1 row affected (0.00 sec)

mysql> select a from t limit 1;
+------+
| a    |
+------+
| utf8 |
+------+
1 row in set (0.00 sec)

mysql> select convert('abc' using (select a from t limit 1));
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(select a from t limit 1))' at line 1
mysql> select convert('abc' using concat('utf', '8'));
ERROR 1115 (42000): Unknown character set: 'concat'
mysql> select convert('abc' using ('utf8'));
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '('utf8'))' at line 1
```
However, if it should be clear to sperate `EvalXXX()` into different stages, I'll try to figure out some other way to  accomplish this.